### ISSUE TITLE
FloppyEmu SmartPort hard disks show as 3.5" floppies on Laser 128

### ISSUE THREAD

**Issue by inexorabletash** [inexorabletash/a2d#6]
SmartPort HDs appear as floppies on Laser 128 w/ Floppy Emu 

No repro on Virtual II (which has its "OmniDisk" smartport drive). Possibilities include:

* A2D is interpreting anything from the Laser's LIRON-ish SmartPort as a floppy
* FloppyEmu is reporting itself incorrectly
* ???

Need to attempt to repro on a non-Laser with LIRON and FloppyEmu and a non-Laser with a IIc or IIgs.

**Comment by inexorabletash** [comment]
A2D is using the low nibble of the unit number (entry in `DEVLST`) to distinguish device types, roughly using:
* If 4, ProFile (if $CnFB = 1) or Slinky RAM disk (otherwise)
* If 0, Disk II (5" floppy)
* If $B, 3.5" floppy
* Otherwise, ProFile

[ProDOS Tech Note #21](http://www.1000bit.it/support/manuali/apple/technotes/pdos/tn.pdos.21.html) says:

> ProDOS 8 does not support an "ID nibble." Section 5.2.4 of the ProDOS 8 Technical Reference Manual states that the low nibble of each unit number in the device list "is a device identification: 0 = Disk II, 4 = Profile, $F = /RAM."
>
> When ProDOS 8 finds a "smart" ProDOS block device while doing its search of the slots and ports, it copies the high nibble of $CnFE (where n is the slot number) into the low nibble of the unit number in the global page. The low nibble then has the following definition:
>
> Bit 3:
>   Medium is removable
> Bit 2:
>   Device is interruptible
> Bit 1-0:
>   Number of volumes on the device (minus one)
>
> As you can see, it is quite easy for the second definition to produce one of the original values (e.g., 0, 4, or $F) in the same nibble for completely different reasons. You should ignore the low nibble in the unit number in the global page when identifying devices since the first definition is insufficient to uniquely identify devices and the second definition contains no information to specifically identify devices. Once you do identify a ProDOS block device, however, you may look at $CnFE to obtain the information in the second definition above, as well as information on reading, writing, formatting, and status availability.

... which means the $B check is really "is removable + 4 volumes", which explains why a SmartPort HD might be detected as such.

Basically... that whole chunk of A2D code needs to follow the TN#21 logic instead. Unfortunately, this is a lot of work as it requires working back from the device list to the driver address to the actual slot to the firmware to a smartport device `STATUS` lookup to get the device type byte per [SmartPort Tech Note #4](http://www.1000bit.it/support/manuali/apple/technotes/smpt/tn.smpt.4.html)

**Comment by inexorabletash** [comment]
Per MG on csa2p, TN#21 doesn't handle ProDOS 2.x behavior. Best thing to do is probably just use size-based heuristics and scrub any code that tries to talk to device firmware.

**Comment by inexorabletash** [comment]
Thoughts on icon heuristics:
* If name starts with `RAM` &rarr; RAMDisk
* &le; 400 blocks &rarr; 140k floppy (normally 280 blocks)
* &le; 2880 blocks &rarr; 800k floppy (handling 400kB, 800kB, 1.44MB)
* else &rarr; ProFile

**Comment by inexorabletash** [comment]
Thanks. I was able to squeeze this into the original space (Main is getting tight, but I see opportunities to dedupe)

* If unit number low nibble is $0, assume Disk II &rarr; 5&frac14;" floppy
* Look for driver address. If not $CnXX &rarr; default
* Look at $CnFF. If $00, Disk II &rarr; 5&frac14;" floppy
* Look for SmartPort signature byte $Cn07. If not $00 &rarr; default
* Look at SmartPort type byte. If low bit set, &rarr; Ram DiskIf name starts with `RAM` &rarr; RAMDisk
* Look at "removable" bit in unit number. If set &rarr; 3&frac12;" floppy
* Else &rarr;  ProFile

This yields the same behavior as before, but is more correct.

**Comment by inexorabletash** [comment]
Okay, previous was due to using ORA instead of AND for masks. D'oh.

And that pointed out a flaw... Virtual II sets its "OmniDisks" to be "Removable" which is sad-making. So... I switched back to just looking at $B for "removable" which is the heuristic DeskTop uses elsewhere, so it's more consistent for now at least. Back to the same old behavior.
